pulsar-broker/src/main/java/org/apache/pulsar/PulsarHealthcheck.java

```diff
@@ -0,0 +1,204 @@
+
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.apache.pulsar.client.admin.PulsarAdminBuilder;
+import org.apache.pulsar.client.api.ClientBuilder;
+import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.Schema;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class PulsarHealthcheck {
+    private static final Logger log = LoggerFactory.getLogger(PulsarHealthcheck.class);
+
+    private static class HealthcheckArguments {
+        @Parameter(names = {"-c", "--clientConf"}, description = "Configuration file for healthcheck")
+        private String healthcheckConf = Paths.get("").toAbsolutePath().normalize().toString() + "/conf/client.conf";
+
+        @Parameter(names = {"-t", "--timeout"}, description = "Max number of seconds to wait for the healthcheck")
+        private int timeoutSeconds = 10;
+
+        @Parameter(names = {"-h", "--help"}, description = "Show this help message")
+        private boolean help = false;
+    }
+
+    private static class Healthcheck implements Callable<Void> {
+        private final Properties properties;
+
+        Healthcheck(Properties properties) {
+            this.properties = properties;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            log.info("Running healthcheck");
+            try (PulsarAdmin admin = propsToAdminBuilder(properties).build()) {
+                String topic = String.format("persistent://%s/healthcheck",
+                                             admin.brokers().getInternalConfigurationData().getHeartbeatNamespace());
+                log.info("Subscribing to topic {}", topic);
+                try (PulsarClient client = propsToClientBuilder(properties).build();
+                     Producer<String> producer = client.newProducer(Schema.STRING).topic(topic).create();
+                     Consumer<String> consumer = client.newConsumer(Schema.STRING)
+                     .subscriptionName("healthcheck-" + UUID.randomUUID().toString()).topic(topic).subscribe()) {
+
+                    String messageStr = UUID.randomUUID().toString();
+                    log.info("Sending message '{}' to {}", messageStr, topic);
+                    producer.send(messageStr);
+                    Message m = consumer.receive(10, TimeUnit.SECONDS);
+                    log.info("Received message '{}' from {}, unsubscribing", m != null ? m.getValue() : "null", topic);
+                    consumer.unsubscribe();
+
+                    if (m == null) {
+                        throw new IllegalStateException("Healthcheck didn't receive message back");
+                    }
+                    if (!m.getValue().equals(messageStr)) {
+                        throw new IllegalStateException("Healthcheck didn't receive back the message it sent");
```

Comment:
are we sure we need this check here? if so, the health check is not able to be used concurrently.

Reply:
This currently can't be used concurrently. I want this check to ensure the publish is actually going through. I can change this to loop receive until we found the one we want.

Reply:
I would suggest changing this to loop receive to improve the health check logic.